src/plugins/janus_sip.c

```diff
@@ -7392,6 +7419,56 @@ gpointer janus_sip_sofia_thread(gpointer user_data) {
 	return NULL;
 }
 
+/* check peer rtp has rfc2833 and push event */
+void janus_sip_check_rfc2833(janus_sip_session *session, char *buffer) {
+	janus_rtp_header *rtp_header = (janus_rtp_header *)buffer;
+	if (rtp_header->type != JANUS_RTP_RFC2833_DEFAULT_PAYLOAD_TYPE)
+		return;
+	char *payload_buffer = buffer + JANUS_RTP_HEADER_LENGTH + rtp_header->csrccount * sizeof(uint32_t);
+	janus_rtp_rfc2833_payload *rfc2833_payload = (janus_rtp_rfc2833_payload *)payload_buffer;
+	uint16_t duration = ntohs(rfc2833_payload->duration);
+	if (rfc2833_payload->end == 0)
+		return;
+
+	/* set up cache to avoid duplication */
+	if(session->dtmf_cache != NULL && session->dtmf_cache->dtmf_event_id == rfc2833_payload->event && session->dtmf_cache->timestamp == rtp_header->timestamp)
+		return;
+	if(session->dtmf_cache != NULL){
+		g_free(session->dtmf_cache);
+		session->dtmf_cache = NULL;
+	}
+	session->dtmf_cache = g_malloc0(sizeof(janus_sip_dtmf_cache));
+	session->dtmf_cache->dtmf_event_id = rfc2833_payload->event;
+	session->dtmf_cache->timestamp = rtp_header->timestamp;
+
+	/* parse dtmf key */ 
+	uint16_t dtmf_key;
+	uint16_t dtmf_keys[] = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '*', '#', 'A', 'B', 'C', 'D'};
+	if (rfc2833_payload->event < 0 || rfc2833_payload->event > 15) {
+  		dtmf_key = '?';
+	} else {
+  		dtmf_key = dtmf_keys[rfc2833_payload->event];
+	}
+	char payload[128];
+	g_snprintf(payload, sizeof(payload), "Signal=%c\nDuration=%d\n", dtmf_key, duration);
+
+	/* Notify the application */
+	json_t *info = json_object();
+	json_object_set_new(info, "sip", json_string("event"));
+	json_t *result = json_object();
+	json_object_set_new(result, "event", json_string("info"));
+	json_object_set_new(result, "sender", json_string(session->callee));
+	json_object_set_new(result, "type", json_string("application/dtmf-relay"));
+	json_object_set_new(result, "content", json_string(payload));
+	if (session->callid)
+		json_object_set_new(info, "call_id", json_string(session->callid));
+	json_object_set_new(info, "result", result);
```

Comment:
I don't like this simulates an info, since this is not a SIP INFO, and may confuse applications into thinking it's supported when it isn't. Better to create a new event (nothing bad about it), e.g., called `dtmf`, where you specify those fields (duration, and key) as separate properties that can be read and parsed much more easily than having to process a crafted string.